File: frontend/src/lib/types/workflow.ts
export interface StageDefinition {
  stage_code: string;
  stage_name: string;
  sequence_order: number;
  sla_days: number;
  statutory_reference: string;
  primary_role: string;
  description: string;
  can_reject: boolean;
  rejection_target?: string;
  required_documents: string[];
}

export interface ProjectStageItem {
  id: string;
  project_id: string;
  stage_code: string;
  stage_name: string;
  sequence_order: number;
  status: 'PENDING' | 'IN_PROGRESS' | 'COMPLETED' | 'BLOCKED' | 'REJECTED';
  compliance_status: 'ON_TRACK' | 'DUE_SOON' | 'OVERDUE' | 'COMPLETED';
  sla_deadline_days: number;
  started_at?: string;
  due_date?: string;
  completed_at?: string;
  days_remaining?: number;
  is_overdue: boolean;
  assigned_role?: string;
  assigned_role_name?: string;
  assigned_user_id?: string;
  assigned_user_name?: string;
  comments?: string;
  rejection_reason?: string;
  required_documents?: string[];
}

export interface TransitionHistoryItem {
  id: string;
  from_stage: string;
  to_stage: string;
  decision: 'APPROVED' | 'REJECTED' | 'OVERRIDDEN';
  remarks?: string;
  triggered_by_user_id: string;
  triggered_by_name?: string;
  triggered_by_role?: string;
  created_at: string;
}

export interface AllowedAction {
  action: 'APPROVED' | 'REJECTED';
  label: string;
  target_stage?: string;
}

export interface ProjectWorkflowTimelineResponse {
  project_id: string;
  project_code: string;
  project_title: string;
  current_stage: string;
  overall_progress_percent: number;
  is_current_stage_overdue: boolean;
  stages: ProjectStageItem[];
  transition_history: TransitionHistoryItem[];
  can_current_user_transition: boolean;
  allowed_transitions: AllowedAction[];
}

export interface StageTransitionRequest {
  decision: 'APPROVED' | 'REJECTED';
  target_stage?: string;
  remarks?: string;
  rejection_reason?: string;
}

export interface WorkflowTaskItem {
  id: string;
  project_id: string;
  project_code?: string;
  project_title?: string;
  parcel_id?: string;
  khasra_number?: string;
  task_type: string;
  title: string;
  description?: string;
  assigned_role: string;
  assigned_role_name?: string;
  assigned_user_id?: string;
  status: 'PENDING' | 'IN_REVIEW' | 'COMPLETED' | 'REJECTED';
  priority: 'NORMAL' | 'HIGH' | 'CRITICAL';
  due_date?: string;
  is_overdue: boolean;
  action_url?: string;
  created_at: string;
}

File: backend/app/schemas/workflow.py
import uuid
from datetime import datetime, date
from typing import List, Optional, Any, Dict
from pydantic import BaseModel, ConfigDict


class StageDefinition(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    stage_code: str
    stage_name: str
    sequence_order: int
    sla_days: int
    statutory_reference: str
    primary_role: str
    description: str
    can_reject: bool = False
    rejection_target: Optional[str] = None
    required_documents: List[str] = []


class ProjectStageItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    stage_code: str
    stage_name: str
    sequence_order: int
    status: str  # PENDING, IN_PROGRESS, COMPLETED, BLOCKED, REJECTED
    compliance_status: str  # ON_TRACK, DUE_SOON, OVERDUE, COMPLETED
    sla_deadline_days: int
    started_at: Optional[datetime] = None
    due_date: Optional[date] = None
    completed_at: Optional[datetime] = None
    days_remaining: Optional[int] = None
    is_overdue: bool = False
    assigned_role: Optional[str] = None
    assigned_role_name: Optional[str] = None
    assigned_user_id: Optional[uuid.UUID] = None
    assigned_user_name: Optional[str] = None
    comments: Optional[str] = None
    rejection_reason: Optional[str] = None
    required_documents: Optional[List[str]] = None


class TransitionHistoryItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    from_stage: str
    to_stage: str
    decision: str
    remarks: Optional[str] = None
    triggered_by_user_id: uuid.UUID
    triggered_by_name: Optional[str] = None
    triggered_by_role: Optional[str] = None
    created_at: datetime


class ProjectWorkflowTimelineResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    project_id: uuid.UUID
    project_code: str
    project_title: str
    current_stage: str
    overall_progress_percent: float
    is_current_stage_overdue: bool
    stages: List[ProjectStageItem]
    transition_history: List[TransitionHistoryItem]
    can_current_user_transition: bool
    allowed_transitions: List[Dict[str, Any]]


class StageTransitionRequest(BaseModel):
    decision: str  # APPROVED, REJECTED
    target_stage: Optional[str] = None
    remarks: Optional[str] = None
    rejection_reason: Optional[str] = None


class WorkflowTaskItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    project_code: Optional[str] = None
    project_title: Optional[str] = None
    parcel_id: Optional[uuid.UUID] = None
    khasra_number: Optional[str] = None
    task_type: str
    title: str
    description: Optional[str] = None
    assigned_role: str
    assigned_role_name: Optional[str] = None
    assigned_user_id: Optional[uuid.UUID] = None
    status: str
    priority: str
    due_date: Optional[date] = None
    is_overdue: bool = False
    action_url: Optional[str] = None
    created_at: datetime


class WorkflowTaskActionRequest(BaseModel):
    action: str  # COMPLETE, REJECT, IN_REVIEW
    remarks: Optional[str] = None
